### PR TITLE
RasterDataset SITS support

### PR DESCRIPTION
**Pull request by sfalkena**
First steps towards Satellite Imagery TimeSeries support within TorchGeo (following #640). Right now I have started with the RasterDataset, but I think my methods can be applied to other GeoDatasets too. I have created a sits_dataset notebook, so feel free to run and play with an example that I prepared.

**Notes**:
- This approach expects the `filename_glob` to be generic enough to pick up on all bands when dealing with `separate_files=True`, so that all files end up in the index. This would lead to the same sampler issues from #2222. If we agree with the general setup of this PR, I will create a separate PR for those sampler fixes. (I have this internally) 
- Currently whenever `return_as_ts=False` shape [c,h,w] is returned, and when `return_as_ts=True` shape [t,c,h,w] is returned. I personally would prefer to move to the [t,c,h,w] standard, and setting `t=1` for non-ts data.
- Next to the SITS, the `__get_item__` also returns the individual dates as an array, so that any model needing the individual dates for the sample can use it.
- Currently, samplers will take the `mint` and `maxt` from the sampler index for timeseries, so that the temporal aspect of the ROI could also limit the temporal aspect. By smartly combining multiple roi's we could then make timeseries for every combination of dates I think.
- `IntersectionDataset` and `UnionDataset` take the `return_as_ts` property based on their input datasets properties.

I have currently not created any unit tests, will do so once we agree on the best structure.

### CODE REVIEW

**Comment by sfalkena** [comment]
Asking explicit feedback from @nilsleh @adamjstewart and @hfangcat

**Comment by adamjstewart** [comment]
I need to find time to sit down and brainstorm this more. May include you and a few others in the discussion. I _think_ the changes proposed here are necessary for what I'm envisioning, but I want to finalize the rest of the picture before I make you do more work. I appreciate your eagerness to work on this, but don't want you to waste too much time on one particular direction before we make a decision.

**Comment by nilsleh** [comment]
@sfalkena Thanks for the contribution and jumping into the effort for time-series support. I am not sure whether you already saw #877 , where we took a stab at this a while ago (and has unfortunately been stale for a while). But I'd be happy to jump in again and brainstorm a good approach for time-series support. In that PR we continued the discussion a bit about possible different use-cases, modalities, and modeling setups worthy of considering, so I'd be interested in what you think about those, just as possible further ideas to consider.

**Comment by sfalkena** [comment]
Have added a new implementation of the SITS dataset, that is a relatively straightforward extension of the current RasterDataset implementation. Added the timeseries parameter to Sentinel dataset for now, will add it to all the other datasets when we agree on the implementation.

Also, feel free to run and plot the SITS dataset by running the tutorial notebook!

**Comment by adamjstewart** [comment]
Can you update the initial comment description? Most of that is out-of-date by now.

File: torchgeo/datasets/geo.py [hunk 1]
```diff
@@ -415,18 +416,24 @@ def __init__(
             transforms: a function/transform that takes an input sample
                 and returns a transformed version
             cache: if True, cache file handle to speed up repeated sampling
+            time_series: if True, return imagery as time series with shape [T,C,H,W]
+                and include 'dates' key with datetime information
```
Comment (adamjstewart) [inline torchgeo/datasets/geo.py]: Let's not refer to it as "imagery", it may be masks or other raster data. Let's just call it "data". 

I'm not sure if "dates" is the right name for the key. It is a datetime, not a date. In other places in TorchGeo, we are using x, y, t for the centroids. We could change that to lon, lat, time.

Let's also add a description like:
"if False, return a merged mosaic with shape [C,H,W]"

File: torchgeo/datasets/geo.py [hunk 2]
```diff
@@ -525,34 +532,89 @@ def __getitem__(self, query: GeoSlice) -> dict[str, Any]:
...
+        if self.time_series:
+            # Group by unique datetime intervals to support timeseries
+            grouped_index = index.groupby(index.index)
+            time_steps = []
+            dates = []
+
+            for datetime_interval, group in grouped_index:
+                # Extract the center date of the interval
+                center_date = datetime_interval.mid
+                dates.append(center_date)
```
Comment (adamjstewart) [inline torchgeo/datasets/geo.py]: I would like to make as many values to be Tensors as we can. In other places, we convert the datetime/pd.Timestamp object to a `.timestamp()` float that can be stored as a Tensor.

File: torchgeo/datasets/geo.py [hunk 3]
```diff
@@ -525,34 +532,89 @@ def __getitem__(self, query: GeoSlice) -> dict[str, Any]:
...
+                    )
+
+                time_steps.append(time_step_data)
+
+            # Stack along time dimension to create [T,C,H,W]
+            data = torch.stack(time_steps, dim=0)
+
         else:
-            data = self._merge_files(index.filepath, query, self.band_indexes)
+            # Original non-timeseries behavior
```
Comment (adamjstewart) [inline torchgeo/datasets/geo.py]: Let's remove this comment, "original" won't make sense in a few years.

File: torchgeo/datasets/geo.py [hunk 4]
```diff
@@ -525,34 +532,89 @@ def __getitem__(self, query: GeoSlice) -> dict[str, Any]:
...
         data = data.to(self.dtype)
         if self.is_image:
             sample['image'] = data
         else:
-            sample['mask'] = data.squeeze(0)
+            # Try to squeeze singleband mask
+            if self.time_series:
+                sample['mask'] = data.squeeze(1)
+            else:
+                sample['mask'] = data.squeeze(0)
```
Comment (adamjstewart) [inline torchgeo/datasets/geo.py]: @ashnair1 do you remember why we added this? In general, I don't like using [un]squeeze, as it can have unintended consequences.

File: torchgeo/datasets/geo.py [hunk 5]
```diff
@@ -525,34 +532,89 @@ def __getitem__(self, query: GeoSlice) -> dict[str, Any]:
...
-                    if match:
-                        if 'band' in match.groupdict():
-                            start = match.start('band')
-                            end = match.end('band')
-                            filename = filename[:start] + band + filename[end:]
-                    filepath = os.path.join(directory, filename)
-                    band_filepaths.append(filepath)
-                data_list.append(self._merge_files(band_filepaths, query))
-            data = torch.cat(data_list)
+        if self.time_series:
```
Comment (adamjstewart) [inline torchgeo/datasets/geo.py:535]: I'm worried that `__getitem__` has become massive. We should use more helper functions here. There is quite a lot of duplicated code that concerns me.

**Review by adamjstewart — CHANGES_REQUESTED** [review]
Only reviewed `geo.py` so far, will review the rest once we finalize `geo.py`

File: torchgeo/datasets/geo.py (hunk 5 above)
Comment (sfalkena) [reply torchgeo/datasets/geo.py:535]: Good point, I have refactored it.

File: torchgeo/datasets/geo.py (hunk 1 above)
Comment (sfalkena) [reply torchgeo/datasets/geo.py]: Renamed "dates" to "timestamps"

File: torchgeo/datasets/geo.py (hunk 2 above)
Comment (sfalkena) [reply torchgeo/datasets/geo.py]: Done

File: torchgeo/datasets/geo.py (hunk 3 above)
Comment (sfalkena) [reply torchgeo/datasets/geo.py]: Done

File: torchgeo/datasets/geo.py (hunk 4 above)
Comment (sfalkena) [reply torchgeo/datasets/geo.py]: My guess was to drop the single channel dimension, but not sure why that was initially in there, I just followed what was there.

File: docs/tutorials/customization.rst [hunk 6]
```diff
@@ -7,6 +7,7 @@ The following tutorials will teach you how to customize TorchGeo to meet your ne
 
 * `Custom Non-Geospatial Datasets <contribute_non_geo_dataset.ipynb>`_: How to create and contribute a new NonGeoDataset
 * `Custom Raster Datasets <custom_raster_dataset.ipynb>`_: How to create a new RasterDataset
+* `SITS Raster Datasets <sits_dataset.ipynb>`_: How to create a RasterDataset with SITS support
```
Comment (adamjstewart) [inline docs/tutorials/customization.rst:10]: Took a quick look at the tutorial. I like it and think it's useful. However, I would put it in Basic Usage. Instead of writing a new dataset, simply show something like:

1. Create a dataset using `Sentinel2(paths)`
2. Create a spatial sampler, show the dimensions it returns
3. Create a new dataset using `Sentinel2(paths, time_series=True)`
4. Create a temporal sampler, show the dimensions it returns

The tests are currently failing because `root` is undefined

File: tests/datasets/test_geo.py [hunk 7]
```diff
@@ -335,20 +336,44 @@ def naip(self, request: SubRequest) -> NAIP:
...
-        return Sentinel2(root, bands=bands, transforms=transforms, cache=cache)
+        time_series = request.param[2]
+        dataset = Sentinel2(
+            root,
+            bands=bands,
+            transforms=transforms,
+            cache=cache,
+            time_series=time_series,
+        )
+        dataset.is_image = request.param[3]
```
Comment (adamjstewart) [inline tests/datasets/test_geo.py:375]: I wouldn't bother testing `is_image` here, unless this is needed to get coverage?

File: torchgeo/datasets/geo.py [hunk 8]
```diff
@@ -234,7 +234,7 @@ def crs(self) -> CRS:
     def crs(self, new_crs: CRS) -> None:
         """Change the :term:`coordinate reference system (CRS)` of a GeoDataset.
 
-        If ``new_crs == self.crs``, does nothing, otherwise updates the index.
+        If ``new_crs == self.crs``, does nothing, otherwise uptimestamps the index.
```
Comment (adamjstewart) [inline torchgeo/datasets/geo.py]:
```suggestion
        If ``new_crs == self.crs``, does nothing, otherwise updates the index.
```

File: torchgeo/datasets/geo.py [hunk 9]
```diff
@@ -525,12 +532,67 @@ def __getitem__(self, query: GeoSlice) -> dict[str, Any]:
...
+
+            # Stack along time dimension to create [T,C,H,W]
+            data = torch.stack(time_steps, dim=0)
+
+        else:
+            data = self._process_files_for_group(index.filepath, query)
+
+        sample: dict[str, Any] = {'crs': self.crs, 'bounds': query}
+        if self.time_series:
+            sample['timestamps'] = torch.tensor(timestamps, dtype=torch.float)
```
Comment (adamjstewart) [inline torchgeo/datasets/geo.py:558]: `timestamps` should be singular just like `image` and `mask`. Not sure if `timestamp` is best, we use `t` in other places but that is less clear.

File: torchgeo/datasets/geo.py [hunk 10]
```diff
@@ -525,12 +532,67 @@ def __getitem__(self, query: GeoSlice) -> dict[str, Any]:
...
+
+        data = data.to(self.dtype)
+        if self.is_image:
+            sample['image'] = data
+        else:
+            # Try to squeeze singleband mask
+            if self.time_series:
+                sample['mask'] = data.squeeze(1)
+            else:
+                sample['mask'] = data.squeeze(0)
```
Comment (adamjstewart) [inline torchgeo/datasets/geo.py]: Can we use -3 instead? That should work for both

File: torchgeo/datasets/geo.py [hunk 11]
```diff
@@ -525,12 +532,67 @@ def __getitem__(self, query: GeoSlice) -> dict[str, Any]:
...
+
+        if self.transforms is not None:
+            sample = self.transforms(sample)
+
+        return sample
+
+    def _process_files_for_group(
```
Comment (adamjstewart) [inline torchgeo/datasets/geo.py]: Name is too generic, this could mean anything

File: torchgeo/datasets/sentinel.py [hunk 12]
```diff
@@ -352,6 +356,8 @@ def __init__(
             transforms: a function/transform that takes an input sample
                 and returns a transformed version
             cache: if True, cache file handle to speed up repeated sampling
+            time_series: if True, return imagery as time series with shape [T,C,H,W]
+                and include 'dates' key with datetime information
```
Comment (adamjstewart) [inline torchgeo/datasets/sentinel.py:360]: Key needs to be renamed here, but let's finalize a name first.

**Review by adamjstewart — CHANGES_REQUESTED** [review]
`__getitem__` logic is still a bit complex. I wonder if we could do something like this. The goal is to get a single data structure that is a list of lists. The inner list represents separate bands that need to be stacked. The outer list represents separate tiles that need to be stacked along a new dimension (if time_series), else merged.

Maybe that's not any better than what you're doing. If I had time, I would try to develop my own solution to see if it becomes just as complex as what you're doing. 

Let's spend more time and use more creativity here and see if there's a simpler way to handle this with fewer if-branches. Right now, it hurts my brain just to look at and try to reason about.

**Comment by adamjstewart** [comment]
Another way to think about this: if we didn't merge files with the same timestamp, I _think_ the code should only require a few lines of changes. So do that first, then see how little we would need to modify to merge files with the same timestamp (possibly in a separate PR).

File: torchgeo/datasets/geo.py (hunk 9 above)
Comment (sfalkena) [reply torchgeo/datasets/geo.py:558]: My opinion is that timestamps is a good description of what it is. However, we could name it timestamp_info, or ts_info if you like?

File: torchgeo/datasets/geo.py (hunk 10 above)
Comment (sfalkena) [reply torchgeo/datasets/geo.py]: Good one, fixed.

File: torchgeo/datasets/geo.py (hunk 8 above)
Comment (sfalkena) [reply torchgeo/datasets/geo.py]: Fixed

File: torchgeo/datasets/geo.py (hunk 11 above)
Comment (sfalkena) [reply torchgeo/datasets/geo.py]: Renamed to `_merge_file_collection`, lmk if you have better suggestions

**Comment by sfalkena** [comment]
I have resolved PR comments. I have given the merging another refactor to make it more functional. Let me know what you think. I will update the tutorial once we agree  on all comments.

File: torchgeo/datasets/geo.py [hunk 13]
```diff
@@ -525,40 +532,106 @@ def __getitem__(self, query: GeoSlice) -> dict[str, Any]:
...
+            ]
+
+        # Process all bands in parallel conceptually
+        band_tensors = [
+            self._merge_files(get_band_filepaths(band), query) for band in self.bands
+        ]
+
+        return torch.cat(band_tensors, dim=0)
+
+    def _construct_band_filepath(
```
Comment (adamjstewart) [inline torchgeo/datasets/geo.py:611]: FYI, we added a new `_update_filepath` function in #3043 similar to this.

File: tests/datasets/test_geo.py (hunk 7 above)
Comment (adamjstewart) [reply tests/datasets/test_geo.py:375]: Ah now I understand, this is indeed needed for coverage.
